sklearn/utils/estimator_checks.py

```diff
@@ -1460,6 +1472,69 @@ def check_estimators_dtypes(name, estimator_orig, strict_mode=True):
                 getattr(estimator, method)(X_train)
 
 
+def check_transformer_preserve_dtypes(
+    name, transformer_orig, strict_mode=True
+):
+    # check that dtype are preserved meaning if input X is of some dtype
+    # X_transformed should be from the same dtype.
+    X, y = make_blobs(
+        n_samples=30,
+        centers=[[0, 0, 0], [1, 1, 1]],
+        random_state=0,
+        cluster_std=0.1,
+    )
+    X = StandardScaler().fit_transform(X)
+    X -= X.min()
+    X = _pairwise_estimator_convert_X(X, transformer_orig)
+
+    def _fit_and_transform(transformer, X, y):
+        transformer = clone(transformer)
+        set_random_state(transformer)
+        X_trans = transformer.fit_transform(X, y)
+
+        if isinstance(X_trans, tuple):
+            # cross-decompostion returns a tuple of (x_scores, y_scores)
+            # when given y with fit_transform; only check the first element
+            X_trans = X_trans[0]
+        return X_trans
+
+    tags_transformer = transformer_orig._get_tags()
+    transformer_default_dtype = tags_transformer["preserves_dtype"][0]
+    X = X.astype(transformer_default_dtype)
+
+    # keep the default dtype as a reference
+    X_trans_default = _fit_and_transform(transformer_orig, X, y)
+    assert X_trans_default.dtype == transformer_default_dtype, (
+            f'Estimator transform dtype: {X_trans_default.dtype} - '
+            f'original/expected dtype: {transformer_default_dtype.__name__}'
+        )
+
+    for dtype in tags_transformer["preserves_dtype"][1:]:
+        X_cast = X.astype(dtype)
+        X_trans = _fit_and_transform(transformer_orig, X_cast, y)
+
+        # check that the output dtype is preserved
+        assert X_trans.dtype == dtype, (
+            f'Estimator transform dtype: {X_trans.dtype} - '
+            f'original/expected dtype: {dtype.__name__}'
+        )
+
+        # check that the transformed array is consistent with the default
+        # transformed array.
+        assert_allclose_dense_sparse(
+            X_trans,
+            X_trans_default,
+            atol=1e-10, rtol=5e-4,
+            err_msg=(
+                f"Transformed array with input dtype {dtype.__name__} is not "
+                f"consistent with reference array of dtype {X.dtype}"
+            )
+        )
```

Comment:
This check is called `check_transformer_preserve_dtypes` and this checks for more than dtype preservation. If a third party estimator has `np.float16` as a preserved dtype, I can see that it would be possible for the `np.float64` result to be different from the `np.float16` result.

Reply:
I agree that checking equality of results could probably be extracted out of this test (and left for another PR). In addition to @thomasjpfan remarks, having consistent results independently of the dtype is something we'd want for all estimators, not just transformers.

Reply:
Agreed. Let's make a core test and we can add array-consistency later on in another test.